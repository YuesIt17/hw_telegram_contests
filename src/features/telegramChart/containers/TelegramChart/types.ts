export type TTelegramChart = {
  width?: number;
  heigth?: number;
}

export type TTelegramChartStyle = Pick<TTelegramChart, 'width'>;
