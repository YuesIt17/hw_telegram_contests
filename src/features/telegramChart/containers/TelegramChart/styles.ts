import {CSSProperties} from 'react';
import {TTelegramChartStyle} from './types';

export const useStyles = ({
  width,
}: TTelegramChartStyle): Record<'chart', CSSProperties | undefined> => ({
  chart: {
    width: width,
    marginTop: 20,
    marginBottom: 20,
    alignSelf: 'flex-start',
  },
});
