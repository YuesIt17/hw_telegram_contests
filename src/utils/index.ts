export {
  prepareData,
  TChartDataLabel,
  TChartDataLine,
  TChartPrepareData,
  TColumnValue,
} from './prepareData';
export {omit} from './utils';
export {
  ThemeTestProvider,
  authTestRender,
  themeTestRender,
  authWrapper,
} from './tests';
