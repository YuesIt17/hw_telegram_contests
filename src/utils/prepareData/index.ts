export {prepareData} from './prepareData';
export {
  TChartDataLabel,
  TChartPrepareData,
  TChartDataLine,
  TColumnValue,
} from './types';
