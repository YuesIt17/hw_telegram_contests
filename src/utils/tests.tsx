import {ThemeProvider} from '@emotion/react';
import {render, RenderOptions, RenderResult} from '@testing-library/react';
import React, {FC, ReactElement} from 'react';
import {theme} from '../app/App';
import {AuthProvider} from '../components/AuthProvider';

export const ThemeTestProvider: FC = ({children}) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export const themeTestRender = (
  ui: ReactElement,
  options?: RenderOptions
): RenderResult => {
  return render(<ThemeTestProvider>{ui}</ThemeTestProvider>, options);
};

export const authTestRender = (
  ui: ReactElement,
  options?: RenderOptions
): RenderResult => {
  return render(
    <AuthProvider>
      <ThemeTestProvider>{ui}</ThemeTestProvider>
    </AuthProvider>,
    options
  );
};
