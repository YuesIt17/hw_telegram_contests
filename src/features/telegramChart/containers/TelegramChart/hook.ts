import {useEffect, useMemo, useReducer} from "react";
import {dataCharts} from "../../../../api/telegramChart";
import {prepareData} from "../../../../utils/prepareData/prepareData";
import {TChartLine} from "../../../../utils/prepareData/types";
import {ChartAction} from "../../store/actions";
import {reducer} from "../../store/reducer";

const initialState: TChartLine[] = [];

export function useTelegramChart() {
  const [state, dispatch] = useReducer(reducer, initialState);
  const {lines, maxDataX, maxDataY, labelsX, labelsY} = useMemo(() => prepareData(dataCharts), [dataCharts]);
  
  useEffect(() => {
    dispatch({type: ChartAction.setAll, payload: lines});
  }, [lines]);
  
  const onChangeHadler = (line: TChartLine, isVisible: boolean) => {
    dispatch({type: ChartAction.updateOne, payload: {...line, isVisible: isVisible}});
  }

  return {data: state, maxDataX, maxDataY, labelsX, labelsY, onChangeHadler}
}
