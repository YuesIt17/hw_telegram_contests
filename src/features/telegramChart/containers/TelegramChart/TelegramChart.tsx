import React from 'react';
import {useStyles} from './styles';
import {TTelegramChart} from './types';
import {useTelegramChart} from './hook';
import {TelegramChartFooter, TelegramChartLine} from '../../components';
import {TelegramChartMap} from '../TelegramChartMap/TelegramChartMap';

export const TelegramChart = ({width}: TTelegramChart) => {
  const styles = useStyles({width});
  const {data, maxDataX, maxDataY, labelsX, labelsY, onChangeHadler} =
    useTelegramChart();

  if (!data) return null;
  return (
    <div style={styles.chart} data-testid="telegramChart">
      <TelegramChartLine
        data={data}
        maxDataX={maxDataX}
        maxDataY={maxDataY}
        labelsX={labelsX}
        labelsY={labelsY}
      />
      <TelegramChartMap
        data={data}
        maxDataX={maxDataX}
        maxDataY={maxDataY}
        width={width}
      />
      <TelegramChartFooter data={data} onChange={onChangeHadler} />
    </div>
  );
};
