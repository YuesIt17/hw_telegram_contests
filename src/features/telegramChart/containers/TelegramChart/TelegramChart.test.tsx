import React from 'react';
import {themeTestRenderer} from '@/utils/tests';
import {TelegramChart} from './TelegramChart';

describe('Telegram Chart test', () => {
  test('Check render TelegramChart component', () => {
    const telegramChart = themeTestRenderer(
      <TelegramChart width={700} />
    ).toJSON();

    expect(telegramChart).toMatchSnapshot();
  });
});
