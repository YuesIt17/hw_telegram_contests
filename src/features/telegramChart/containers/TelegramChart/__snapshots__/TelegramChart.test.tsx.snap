// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telegram Chart test Check render TelegramChart component 1`] = `
<div
  className="css-1km59tu-chart"
  data-testid="telegramChart"
>
  <svg
    style={
      Object {
        "width": undefined,
      }
    }
    viewBox="0 0 591 308"
  >
    <g>
      <polyline
        className="css-6b3m9l-line"
        points="0,278 561,278"
      />
      <text
        className="css-1xdcvwp-label"
        x={1}
        y={302}
      >
        Nov 17
      </text>
      <text
        className="css-1xdcvwp-label"
        x={61}
        y={302}
      >
        Nov 29
      </text>
      <text
        className="css-1xdcvwp-label"
        x={121}
        y={302}
      >
        Dec 11
      </text>
      <text
        className="css-1xdcvwp-label"
        x={181}
        y={302}
      >
        Dec 23
      </text>
      <text
        className="css-1xdcvwp-label"
        x={241}
        y={302}
      >
        Jan 4
      </text>
      <text
        className="css-1xdcvwp-label"
        x={301}
        y={302}
      >
        Jan 16
      </text>
      <text
        className="css-1xdcvwp-label"
        x={361}
        y={302}
      >
        Jan 28
      </text>
      <text
        className="css-1xdcvwp-label"
        x={421}
        y={302}
      >
        Feb 9
      </text>
      <text
        className="css-1xdcvwp-label"
        x={481}
        y={302}
      >
        Feb 21
      </text>
      <text
        className="css-1xdcvwp-label"
        x={541}
        y={302}
      >
        Mar 5
      </text>
    </g>
    <g>
      <polyline
        className="css-6b3m9l-line"
        points="0,0 0,278"
      />
      <g>
        <text
          className="css-1xdcvwp-label"
          x={0}
          y={206}
        >
          60
        </text>
        <polyline
          className="css-6b3m9l-line"
          points="0,218 561,218"
        />
      </g>
      <g>
        <text
          className="css-1xdcvwp-label"
          x={0}
          y={146}
        >
          120
        </text>
        <polyline
          className="css-6b3m9l-line"
          points="0,158 561,158"
        />
      </g>
      <g>
        <text
          className="css-1xdcvwp-label"
          x={0}
          y={86}
        >
          180
        </text>
        <polyline
          className="css-6b3m9l-line"
          points="0,98 561,98"
        />
      </g>
      <g>
        <text
          className="css-1xdcvwp-label"
          x={0}
          y={26}
        >
          240
        </text>
        <polyline
          className="css-6b3m9l-line"
          points="0,38 561,38"
        />
      </g>
    </g>
  </svg>
  <div
    className="css-x43oq9-chartMap"
    data-testid="telegramChartMap"
  >
    <svg
      style={
        Object {
          "width": undefined,
        }
      }
      viewBox="0 0 561 56"
    />
  </div>
  <div
    className="css-1weht41-footer"
  />
</div>
`;
